lua_modules/ftp/ftpserver.lua

```diff
@@ -0,0 +1,163 @@
+-- a simple ftp server
+local file,net,pairs,print,string,table = file,net,pairs,print,string,table
+local ftp, ftp_data, ftp_srv
+do
+local createServer = function (user, pass)
+local data_fnc = nil
+ftp_data = net.createServer(net.TCP, 180)
+ftp_data:listen(20, function (s) if data_fnc then data_fnc(s) end end)
+ftp_srv = net.createServer(net.TCP, 180)
+ftp_srv:listen(21, function(socket)
+  local s = 0
+  local cwd = "/"
+  local buf = ""
+  local t = 0
+  socket:on("receive", function(c, d)
+    local a = {}
+    for i in string.gmatch(d, "([^ \r\n]+)") do
+      table.insert(a,i)
+    end
+    local a1,a2 = unpack(a)
+    if a1 == nil or a1 == "" then return end
+    if s == 0 and a1 == "USER" then
+      if a2 ~= user then
+        return c:send("530 user not found\r\n")
+      end
+      s = 1
+      return c:send("331 OK. Password required\r\n")
+    end
+    if s == 1 and a1 == "PASS" then
+      if a2 ~= pass then
+        return c:send("530 Try again\r\n")
+      end
+      s = 2
+      return c:send("230 OK.\r\n")
+    end
+    if s ~= 2 then
+      return c:send("530 Not logged in, authorization required\r\n")
+    end
+    if a1 == "CDUP" then
+      return c:send("250 OK. Current directory is "..cwd.."\r\n")
+    end
+    if a1 == "CWD" then
+      if a2 == "." then
+        return c:send("257 \""..cwd.."\" is your current directory\r\n")
+      end
+      cwd = a2
+      return c:send("250 OK. Current directory is "..cwd.."\r\n")
+    end
+    if a1 == "PWD" then
+      return c:send("257 \""..cwd.."\" is your current directory\r\n")
+    end
+    if a1 == "TYPE" then
+      if a2 == "A" then
+        t = 0
+        return c:send("200 TYPE is now ASII\r\n")
+      end
+      if a2 == "I" then
+        t = 1
+        return c:send("200 TYPE is now 8-bit binary\r\n")
+      end
+      return c:send("504 Unknown TYPE")
+    end
+    if a1 == "MODE" then
+      if a2 ~= "S" then
+        return c:send("504 Only S(tream) is suported\r\n")
+      end
+      return c:send("200 S OK\r\n")
+    end
+    if a1 == "PASV" then
+      local _,ip = socket:getaddr()
+      local _,_,i1,i2,i3,i4 = string.find(ip,"(%d+).(%d+).(%d+).(%d+)")
+      return c:send("227 Entering Passive Mode ("..i1..","..i2..","..i3..","..i4..",0,20).\r\n")
+    end
+    if a1 == "LIST" or a1 == "NLST" then
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        local l = file.list();
+        for k,v in pairs(l) do
+          sd:send( a1 == "NLST" and k.."\r\n" or "+r,s"..v..",\t"..k.."\r\n")
+        end
+        sd:close()
+        data_fnc = nil
+        c:send("226 Transfer complete.\r\n")
+      end
+      return
```

Comment:
Proposal for "early connect" scenario:
```lua
if data_sock then
  node.task.post(function() data_fnc(data_sock); data_sock = nil end)
end
return
```
This starts `data_fnc` manually after the return.